pkg/storage/secret/metadata/inline_store.go

```diff
@@ -0,0 +1,147 @@
+package metadata
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	authlib "github.com/grafana/authlib/types"
+	common "github.com/grafana/grafana/pkg/apimachinery/apis/common/v0alpha1"
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	secret "github.com/grafana/grafana/pkg/apis/secret/v0alpha1"
+	"github.com/grafana/grafana/pkg/registry/apis/secret/contracts"
+	"github.com/grafana/grafana/pkg/registry/apis/secret/xkube"
+)
+
+var _ contracts.InlineSecureValueSupport = (*inlineStorage)(nil)
+
+func ProvideInlineSecureValueSupport(
+	db contracts.SecureValueMetadataStorage,
+	access authlib.AccessClient,
+) (contracts.InlineSecureValueSupport, error) {
+	return &inlineStorage{
+		db:     db,
+		access: access,
+	}, nil
+}
+
+// TODO! this needs real attention :) and likely explicit support from the DB
+// inlineStorage is the actual implementation of the secure value (metadata) storage.
+type inlineStorage struct {
+	db     contracts.SecureValueMetadataStorage
+	access authlib.AccessChecker
+}
+
+// CanReference implements contracts.InlineSecureValueSupport.
+func (i *inlineStorage) CanReference(ctx context.Context, owner common.ObjectReference, values common.InlineSecureValues) (bool, error) {
+	actor, ok := authlib.AuthInfoFrom(ctx)
+	if !ok {
+		return false, apierrors.NewBadRequest("missing auth info")
+	}
+	if owner.Name == "" {
+		return false, apierrors.NewBadRequest("missing owner name")
+	}
+
+	// TODO? more efficient version of this based on database query?
+	// We want to check if the owner matches OR the identity in context have view permissions
+	for _, value := range values {
+		if !value.Create.IsZero() {
+			return false, apierrors.NewBadRequest("unable to create")
+		}
+		if value.Remove {
+			return false, apierrors.NewBadRequest("unable to remove")
+		}
+		if value.Name == "" {
+			return false, apierrors.NewBadRequest("expecting a name")
+		}
+
+		// ???? Read requires view?  do we have a read that does not require view
+		v, err := i.db.Read(ctx, xkube.Namespace(owner.Namespace), value.Name, contracts.ReadOpts{})
```

Comment:
bit confused about this piece -- should we be failing here, before we have checked if the owner references match?

Reply:
The *real* implementation will need to do that -- I think the db interface may need updated to make this more friendly/efficient.